src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/RepositoryCloneManager.cs

```diff
@@ -30,66 +36,112 @@ public class RepositoryCloneManager : IRepositoryCloneManager
     private readonly IVmrInfo _vmrInfo;
     private readonly ILocalGitRepo _localGitRepo;
     private readonly IGitRepoClonerFactory _remoteFactory;
-    private readonly IProcessManager _processManager;
     private readonly IFileSystem _fileSystem;
     private readonly ILogger<VmrPatchHandler> _logger;
 
+    // Map of URI => dir name
+    private readonly Dictionary<string, LocalPath> _clones = new();
+
+    // Repos we have already pulled updates for during this run
     private readonly List<string> _upToDateRepos = new();
 
     public RepositoryCloneManager(
         IVmrInfo vmrInfo,
         ILocalGitRepo localGitRepo,
         IGitRepoClonerFactory remoteFactory,
-        IProcessManager processManager,
         IFileSystem fileSystem,
         ILogger<VmrPatchHandler> logger)
     {
         _vmrInfo = vmrInfo;
         _localGitRepo = localGitRepo;
         _remoteFactory = remoteFactory;
-        _processManager = processManager;
         _fileSystem = fileSystem;
         _logger = logger;
     }
 
-    public async Task<LocalPath> PrepareClone(string repoUri, string checkoutRef, CancellationToken cancellationToken)
+    public LocalPath PrepareClone(
+        SourceMapping mapping,
+        string[] remoteUris,
+        string checkoutRef,
+        CancellationToken cancellationToken)
     {
-        cancellationToken.ThrowIfCancellationRequested();
+        if (remoteUris.Length == 0)
+        {
+            throw new ArgumentException("No remote URIs provided to clone");
+        }
+
+        LocalPath path = null!;
+        bool needsCheckout = false;
+        foreach (string remoteUri in remoteUris)
+        {
+            // Path should be returned the same for all invocations
+            (path, needsCheckout) = PrepareClone(remoteUri, checkoutRef, mapping.Name, cancellationToken);
+        }
+
+        if (needsCheckout)
+        {
+            _localGitRepo.Checkout(path, checkoutRef); 
+        }
 
-        var hash = GetCloneDirName(repoUri);
-        var clonePath = _vmrInfo.TmpPath / hash;
+        return path;
+    }
+
+    public LocalPath PrepareClone(
+        string repoUri,
+        string checkoutRef,
+        CancellationToken cancellationToken)
+    {
+        var result = PrepareClone(repoUri, checkoutRef, GetHash(repoUri), cancellationToken);
```

Comment:
so if there is only one URL it's hashed, but if there are many, the clone folder is the name of the mapping?

Reply:
If you supply a mapping, the folder will be the mapping name, yes. If you later add a new remote to a mapping, it will be added there as well. If you are cloning submodules by URL directly, it will be a hash because we don't have a name for those.
If you cloned by a mapping first and later you are cloning by URL (e.g. for a VMR patch name), the URL will be looked up with the mapping too.

Reply:
I see, thanks